frame/democracy/src/lib.rs

```diff
@@ -525,34 +594,57 @@ decl_module! {
 
 		fn deposit_event() = default;
 
+		fn on_runtime_upgrade() -> Weight {
+			if StorageVersion::get() == Releases::V1 {
+				DepositOf::<T>::translate::<
+					(BalanceOf<T>, Vec<T::AccountId>), _
+				>(|_, (balance, accounts)| {
+					Some((accounts, balance))
+				});
+
+				StorageVersion::put(Releases::V2);
+
+				T::MaximumBlockWeight::get()
+			} else {
+				T::DbWeight::get().reads(1)
+			}
+		}
+
 		/// Propose a sensitive action to be taken.
 		///
 		/// The dispatch origin of this call must be _Signed_ and the sender must
 		/// have funds to cover the deposit.
 		///
 		/// - `proposal_hash`: The hash of the proposal preimage.
 		/// - `value`: The amount of deposit (must be at least `MinimumDeposit`).
+		/// - `proposals_upper_bound`: an upper bound on the current number of public proposals.
+		///   Extrinsic is weighted according to this value with no refund.
 		///
 		/// Emits `Proposed`.
 		///
 		/// # <weight>
-		/// - `O(P)`
-		/// - P is the number proposals in the `PublicProps` vec.
-		/// - Two DB changes, one DB entry.
+		/// - Complexity: `O(P)` where `P` is the number of proposals in the `PublicProps` vec.
+		/// - Db reads: `PublicPropCount`, `PublicProps`
+		/// - Db writes: `PublicPropCount`, `PublicProps`, `DepositOf`
 		/// # </weight>
-		#[weight = 5_000_000_000]
+		#[weight = 42_000_000
+			.saturating_add(220_000.saturating_mul(Weight::from(*proposals_upper_bound)))
+			.saturating_add(T::DbWeight::get().reads_writes(2, 3))
+		]
 		fn propose(origin,
 			proposal_hash: T::Hash,
-			#[compact] value: BalanceOf<T>
+			#[compact] value: BalanceOf<T>,
+			#[compact] proposals_upper_bound: u32,
 		) {
 			let who = ensure_signed(origin)?;
 			ensure!(value >= T::MinimumDeposit::get(), Error::<T>::ValueLow);
 			T::Currency::reserve(&who, value)?;
 
 			let index = Self::public_prop_count();
 			PublicPropCount::put(index + 1);
-			<DepositOf<T>>::insert(index, (value, &[&who][..]));
+			<DepositOf<T>>::insert(index, (&[&who][..], value));
 
+			ensure!(Self::len_of_public_props() <= proposals_upper_bound, Error::<T>::WrongUpperBound);
```

Comment:
Verify first!!!!!!!!!!!

please write a test

Reply:
damn sorry, I do so

Reply:
Would it be a general thing? I don't see it is a bad idea, but it is also a little bit ugly.

Hard for me to say if this is a good idea or not...

Reply:
I mean it can be written like this:
```rust
fn my_extrinsic(origin, ...) -> DispatchResult {
  // Do stuff and can return errors
  with_mutable(|stg| -> () { // This closure can't return
     MyStorage::put(stg, 3u32);
  })
}
```
But yes passing around stg is a bit anoying.